RemovedPackages: added noauto postupgrade script

pirat89 opened the pull request:
In case user doesn't use --cleanup-post option in r-u-t, there could be kept broken el6 packages which could inhibit user to update their package set. This script can be run by user manually to resolve such issues, keeping non-native packages untouched.

pirat89 commented:
@AloisMahdal In this case, I think just partial tests are required. I guess we do not have tests for this module specifically, do we? Currenly expected behaviour:
- expected files are copied under the noauto_postupgrade.d dir    
- the script remove all old native packages with broken dependencies    
- custom rpms are by default kept untouched (doesn't matter whether deps are broken or not here)
  - when the script is called with parameter `--rm-all`, remove even non-native packages when they have broken deps
- in case some new-native packages have broken dependencies, install dependencies when they are available
- in case something is wrong, inform user that manual steps are required       
     
- in case a "branch" of tree with broken dependencies is longer than 10 rpms, probably some packages have not been still removed and script has to be run again; in that case, user is informed whether some rpms with broken deps remain and should be removed.
  - Example: let's have 15 rpms with cascade deps (rpm15 requires rpm14, ...) and only rpm1 has now broken dependency. By removing of rpm1 we break deps of rpm2 (nothing else). Remove rpm2 in another iteration so rpm3 has broken deps, ....

AloisMahdal commented on RHEL6_7/packages/RemovedPackages/noauto_postupgrade.d/native_rpm_dep_fix.sh:
```diff
@@ -22,22 +29,33 @@ is_dist_native() {
+    #
+    # Remove native RPMs with broken dependencies.
+    #
+    # The fce can be affected by the $RM_CUSTOM_RPM variable to remove even
```
"fce" is a contraction of Czech "funkce".

pirat89 replied:
Thanks :-)

pirat89 replied:
fixed

pirat89 commented:
Rebased for 6.10 -> 7.5 upgrade path.

pirat89 commented:
--test--

jmazanek commented on RHEL6_7/packages/RemovedPackages/noauto_postupgrade.d/native_rpm_dep_fix.sh:
```diff
@@ -99,7 +103,7 @@ for counter in {0..10}; do
+    rm_broken_old_rpms "$RM_CUSTOM_RPM" && break
     [ $counter -eq 10 ] && {
         [ -z "$(get_broken_deps_list)" ] && {
             log_warning "Reached iteration limit but some RPMs are still broken."
```
The log warning does not tell the user, how to resolve the situation. Add there the information about, what should user do, when it happens (re-run the script).

pirat89 replied:
it is written just line below :)

bocekm commented on RHEL6_7/packages/RemovedPackages/check:
```diff
@@ -68,8 +85,9 @@ echo \
+
 " >>solution.txt
-[ $found -eq 1 ] && log_high_risk "After upgrading to Red Hat Enterprise Linux 7, there are still some el6 packages left. Add the '--cleanup-post' option to redhat-upgrade-tool to remove them automatically."
+[ $found -eq 1 ] && log_high_risk "After upgrading to Red Hat Enterprise Linux 7, there are still some el6 packages left. Add the '--cleanup-post' option to redhat-upgrade-tool to remove them automatically. You can also run manually the $NOAUTO_POST_DIR/$NOAUTO_SCRIPT_NAME script after the upgrade to remove only old native packages with broken dependencies."
```
- *Red Hat Enterprise Linux 7, there are still some el6* **->** *Red Hat Enterprise Linux (RHEL) 7, there are still some RHEL 6*
- *run manually the XY script* **->** *run the XY script manually*
- *only old native packages with broken dependencies* **->** *only those Red Hat-signed RHEL 6 packages that have broken dependencies*

pirat89 replied:
fixed

bocekm commented on RHEL6_7/packages/RemovedPackages/check:
```diff
@@ -32,6 +43,12 @@ echo \
 
+Use the '--cleanup-post' option of redhat-upgrade-tool to remove them
+automatically during upgrade. You can also run manually the $NOAUTO_POST_DIR/$NOAUTO_SCRIPT_NAME script
+after the upgrade to remove only the original native packages with broken
```
- *run manually the XY script* **->** *run the XY script manually*
- *only old native packages with broken dependencies* **->** *only those Red Hat-signed RHEL 6 packages that have broken dependencies*

pirat89 replied:
fixed

bocekm commented on RHEL6_7/packages/RemovedPackages/noauto_postupgrade.d/native_rpm_dep_fix.sh:
```diff
@@ -0,0 +1,135 @@
+        is_dist_native "$NAME" || [ $rm_custom_enabled -eq 1 ] || {
+            # it's not dist native package -> skip it
+            echo "    $NAME" >> "$kept_broken_rpms"
+            log_info_verbose "Skip the non-native $line RPM."
```
_non-native_ **->** *3rd party or custom*

pirat89 replied:
fixed

bocekm commented on RHEL6_7/packages/RemovedPackages/noauto_postupgrade.d/native_rpm_dep_fix.sh:
```diff
@@ -0,0 +1,135 @@
+            ;;
+        -h | --help)
+            echo "USAGE: native_rpm_dep_fix [-a|--rm-all] [-v|--verbose] [-h|--help]"
+            echo "    -a | --rm-all   Enable remove of non-native RPMs"
```
_remove of non-native_ **->** *removal of 3rd party or custom*

pirat89 replied:
fixed

bocekm commented on RHEL6_7/packages/RemovedPackages/noauto_postupgrade.d/native_rpm_dep_fix.sh:
```diff
@@ -0,0 +1,135 @@
+touch "$kept_broken_rpms"
+
+[ $RM_CUSTOM_RPM -eq 1 ] \
+    && log_info_verbose "Remove of non-native conflicting RPMs has been enabled."
```
_Remove of non-native_ **->** ***Removal** of 3rd party or custom*
BTW, you're using "conflicting" here for the first and only time. I would use what you use on the other places - "with broken dependencies".

pirat89 replied:
You are right. Fixed.

bocekm commented on RHEL6_7/packages/RemovedPackages/noauto_postupgrade.d/native_rpm_dep_fix.sh:
```diff
@@ -0,0 +1,135 @@
+broken_nonnative_rpms=$(wc -l <"$kept_broken_rpms")
+[ $broken_nonnative_rpms -gt 0 ] && {
+    # this is obviously irrelevant in case of use the -a option
+    log_warning "Detected several non-native RPMs with broken dependencies:"
```
*non-native* **->** *3rd party or custom RHEL 6*

pirat89 replied:
fixed

AloisMahdal commented:
There was a discussion about this in Bugzilla and/or RH mailing list long time ago.  The "native" term was intentionally introduced to distinguish packages that are owned by distribution from others.  I also recommended to avoid term "3rd party" because it IMHO can be confusing if the "3rd party software" is actually one written by customer--so there's no 3rd party; there's just distribution SW and user SW.  Note that this *is* the only distinction that can ever matter to us: we are'nt supposed to care/imply who maintains the packages if it's not *us*.

Whether `el`, `rhel` or `el6` should be used here depends on whether or not the module is RHEL(6)-specific.  If it inherently is, then it's OK to hardcode `el6` in the name, if not, it's IMO best to avoid.

bocekm commented:
@AloisMahdal that is nice that you guys introduced it, but it's not documented anywhere and customers will be left wondering what exactly that means. So as I mentioned in the review, I would either rephrase it to avoid the term or explain the term. I agree with you that when using 3rdparty and custom, it's again using terms that are not clearly defined in our project but in my eyes are more self-explanatory. Anyway, I'll reach out to the docs team to ask them what way to go.

AloisMahdal commented:
Sorry, but we won't make it in this release; there's been too many other issues.  I talked to @pirat89 and we've agreed we can move it to the next  one.

AloisMahdal commented on RHEL6_7/packages/RemovedPackages/noauto_postupgrade.d/native_rpm_dep_fix.sh:
```diff
@@ -0,0 +1,135 @@
+
+log_error()   { echo >&2 "Error:   $@"; }
+log_warning() { echo >&2 "Warning: $@"; }
+log_info()    { echo >&2 "Info:    $@"; }
```
Mixing array and string.  Did you mean this?

    log_error() { echo >&2 "Error: " "$@"; }

or (argument is line):

    log_error() { for msg in "$@"; do echo >&2 "Error: $msg"; done; }

or just ignore from $2 on::

    log_error() { echo >&2 "Error: $2"; }

pirat89 replied:
The forth option :D :
```
log_error() { echo >&2 "Error: $1"; }
```

AloisMahdal commented on RHEL6_7/packages/RemovedPackages/noauto_postupgrade.d/native_rpm_dep_fix.sh:
```diff
@@ -0,0 +1,135 @@
+
+#RHSIGNED_PKGS="/var/cache/preupgrade/common/rpm_rhsigned.log"
+RHSIGNED_PKGS="rpm_rhsigned.log"
+kept_broken_rpms="kept_broken_rpms"
```
I'm not sure why you need variable if the value is always just the variable name itself...

pirat89 replied:
I have not been sure that there will not be different name later (because of *broken*) and have not a mood to think about something better. Anyway, yes, it's not neede, but I havn't wanted to correct that several times in case there will be different name used later. And shellckeck works better when you uses variables to detect possible typos :)

AloisMahdal commented on RHEL6_7/packages/RemovedPackages/noauto_postupgrade.d/native_rpm_dep_fix.sh:
```diff
@@ -0,0 +1,135 @@
+[ $broken_nonnative_rpms -gt 0 ] && {
+    # this is obviously irrelevant in case of use the -a option
+    log_warning "Detected several 3rd party or custom RHEL 6 RPMs with broken dependencies:"
+    cat "$kept_broken_rpms" | sort | uniq >&2
```
Useless use of cat :)

    <"$stuff" sort | uniq >&2

pirat89 replied:
I know. But this is only cat I really like, so I put it on many places :heart:

AloisMahdal commented on RHEL6_7/packages/RemovedPackages/noauto_postupgrade.d/native_rpm_dep_fix.sh:
```diff
@@ -0,0 +1,135 @@
+            VERBOSE=1
+            ;;
+        -h | --help)
+            echo "USAGE: native_rpm_dep_fix [-a|--rm-all] [-v|--verbose] [-h|--help]"
```
`native_rpm_dep_fix` is not true anymore.  You could use `${0##*/}` here, I think.

pirat89 replied:
right! will be fixed in the next commit.

AloisMahdal commented:
Checked few basic cases and the script works fine.  However, we don't have time anymore for proper verification (which is also why this was not acked.)

pirat89 commented:
@tcerna @AloisMahdal @drehak: Fixed some reported issues and rebased with master. All new changes are part of the last commit. Please, do a review + give an ack (in case you will not find an issue inside the changes :) ) and I will merge it.

pirat89 commented:
Moved to the next milestone for proper testing.

pirat89 commented:
Removing from the milestone as stil it is not tested.

pirat89 commented:
We will celebrate 1y soon :-P

pirat89 commented:
-- build --

pirat89 commented:
--test--

pirat89 commented:
@bocekm @AloisMahdal what about this *nestor*?

AloisMahdal commented:
> @bocekm @AloisMahdal what about this _nestor_?

No idea; if it's worth we can try and test it again but it's not like we have extra resources... :-/

pirat89 commented:
Keep in mind it's manual script only - it will not be executed automatically. I guess that manual testing should not take so much time and it could be tested by anyone.

AloisMahdal commented:
> Keep in mind it's manual script only - it will not be executed automatically

I don't see how this is relevant.


> should not take so much time 

"should not"?  or more like "I wish it would not"? :-)

Looking at the spec, there's actually lot of test cases and some of them would not be easy to set up.

pirat89 commented:
So let's make it simple. I am closing the PR as it doesn't look like it will be merged during next 2 working days. People will have to deal with the problems on their own. If they will not, we can point them that they can download the script manually from github.
